### PR TITLE
ci: update CD process

### PR DESCRIPTION
**Pull request by deantchi**
changes to the CI workflow.
adding auto deployments using docker container digest to the k8s repo.
dev and stg will auto approve, while prod will require manual approval

### CODE REVIEW

File: .github/workflows/ci.yaml
Comment (CharlieC3) [inline .github/workflows/ci.yaml:89]: These 3 envs need to have required reviewers configured, otherwise it'll just deploy to them without an approval. [You can configure that here.](https://github.com/hirosystems/chainhook/settings/environments)

File: .github/workflows/ci.yaml
Comment (CharlieC3) [inline .github/workflows/ci.yaml:101]: I don't believe the build-publish job is outputting anything, so this will fail. Double check any other outputs used too.

File: .github/workflows/ci.yaml
Comment (CharlieC3) [inline .github/workflows/ci.yaml:165]: This doesn't seem right, we don't want to allow deployments to prod except under more rigid conditions. Double check this condition with the workflow in the [stacks-devnet-api](https://github.com/hirosystems/stacks-devnet-api/blob/develop/.github/workflows/ci.yaml#L287).

**Review by CharlieC3 — CHANGES_REQUESTED** [review]
Good start! Just a few tweaks.

File: .github/workflows/ci.yaml
Comment (deantchi) [reply .github/workflows/ci.yaml:89]: opps yeah, I added the reviewers but forgot to save the changes

**Comment by CharlieC3** [comment]
@deantchi you'll also need to update the chainhook manifests to apply the environment-level patch using `digest` instead of `newTag`.
See the Explorer or stacks-devnet-api manifests for a reference if needed.

File: .github/workflows/ci.yaml
Comment (CharlieC3) [inline .github/workflows/ci.yaml:184]: We should take testnet into account as well I think.

Comment (deantchi) [reply .github/workflows/ci.yaml]: trying out the use of matrix strategy for testnet | mainnet

Comment (deantchi) [reply .github/workflows/ci.yaml]: this seems to be working.  https://github.com/hirosystems/chainhook/actions/runs/6356417456

File: .github/workflows/ci.yaml
Comment (CharlieC3) [inline .github/workflows/ci.yaml:181]:
```suggestion
    if: needs.build-publish.outputs.new_release_published == 'true' && (github.event_name != 'pull_request' || github.event.pull_request.head.repo.full_name == github.repository)
```

File: .github/workflows/ci.yaml
Comment (CharlieC3) [inline .github/workflows/ci.yaml:88]: I like what you did with the matrix! However on the approval screen, it doesn't seem like it's going to let us differentiate between a testnet dev deployment and a mainnet dev deployment. So I think they only way to get around this is to create a total of 6 Github Environments, aptly named something like Development - Mainnet, Development - Testnet, etc. You might still be able to leverage the matrix usage here as well.

**Review by CharlieC3 — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yaml
Comment (CharlieC3) [reply .github/workflows/ci.yaml:88]: Despite there being two deployments prepped, there's only one approval box. Ideally devs would have control to deploy to both networks for an environment, or just one.
Screenshot 2023-09-29 at 8 39 30 PM

**Comment by lgalabru** [comment]
🤤 
Thank you for this new flow @deantchi @CharlieC3!

**Review by MicaiahReid — APPROVED** [review]

File: .github/workflows/ci.yaml
Comment (deantchi) [reply .github/workflows/ci.yaml:88]: created 6 separate deployment environments as suggested and using matrix function

**Review by CharlieC3 — APPROVED** [review]
I made a few changes to help with CI performance, specifically added caching for cargo tests and docker image build. In addition to that, I fixed some dependency issues between jobs, ensured the new tag is being checked out once a new release is published, removed the `blockstack` Docker hub org from being used, and some other minor fixes.

I'd like either @MicaiahReid or @deantchi to check this PR again and re-approve if possible.

**Pull request merged**
